### ISSUE TITLE
` pytorch` learning rate too aggressive on my dataset (defaults with TF are fine)

### ISSUE THREAD

**Issue by tsievert** [DeepLabCut/DeepLabCut#2922]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Bug description

I have a model that I trained for 5 iteration with great succes. I labelled extra outliers and upgraded to 3.0.0rc7 and wanted to train the new iteration with pytorch.
The pytorch training shows basically now learning, but if I switch back to TF, the model seems to learn just fine.
I have about 2300 labelled frames, in case larger datasets are probelematic.

### Operating System

operating systVersion: Linux version 4.18.0-553.34.1.el8_10.x86_64 ([mockbuild@x86-64-03.build.eng.rdu2.redhat.com](mailto:mockbuild@x86-64-03.build.eng.rdu2.redhat.com)) (gcc version 8.5.0 20210514 (Red Hat 8.5.0-22) (GCC)) 

### DeepLabCut version

3.0.0rc7

### DeepLabCut mode

single animal

### Device type

gpu

### Steps To Reproduce

1. Have existing trained model (pre DLC v3.0.0)
2. Extract outliers
3. Upgrade to latest 3.0.0rc
4. Label frames
5. Switch `config.yaml` to pytorch
6. Train model for 500 epochs
7. Observe no learning
```bash
Evaluation results for DLC_Resnet50_Callibration_analysisSep14shuffle1_snapshot_010-results.csv (pcutoff: 0.8):
train rmse            933.00
train rmse_pcutoff       NaN
train mAP               0.00
train mAR               0.00
test rmse             934.06
test rmse_pcutoff        NaN
test mAP                0.00
test mAR                0.00
Name: (0.95, 1, 10, -1, 0.8), dtype: float64
```
8. Add the engine switch to TF, train on same data
9. Normal learning

[learning_stats.csv](https://github.com/user-attachments/files/19234053/learning_stats.csv)

### Relevant log output

```shell

```

### Anything else?

I can provide all or a subset of my labelled frames if that helps.

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by n-poulsen** [comment]
Hi @tsievert! Thanks for reporting this - could you share the content of your `pytorch_config.yaml` with me?

**Comment by tsievert** [comment]
Sure, here you go:
```yaml
data:
  colormode: RGB
  inference:
    normalize_images: true
  train:
    affine:
      p: 0.5
      rotation: 30
      scaling:
      - 0.5
      - 1.25
      translation: 0
    covering: false
    crop_sampling:
      width: 448
      height: 448
      max_shift: 0.1
      method: hybrid
    gaussian_noise: 12.75
    hist_eq: false
    motion_blur: false
    normalize_images: true
device: auto
metadata:
  project_path: 
    /gpfs/projects/acad/behavior/DLC_model/primary_DeepLabCut_model/LassanceLab-main-model-2023-09-14
  pose_config_path: 
    /gpfs/projects/acad/behavior/DLC_model/primary_DeepLabCut_model/LassanceLab-main-model-2023-09-14/dlc-models-pytorch/iteration-7/Callibration_analysisSep14-trainset95shuffle1/train/pytorch_config.yaml
  bodyparts:
  - nose
  - left_ear
  - right_ear
  - left_ear_tip
  - right_ear_tip
  - left_eye
  - right_eye
  - neck
  - mid_back
  - mouse_center
  - mid_backend
  - mid_backend2
  - mid_backend3
  - tail_base
  - tail1
  - tail2
  - tail3
  - tail4
  - tail5
  - left_shoulder
  - left_midside
  - left_hip
  - right_shoulder
  - right_midside
  - right_hip
  - tail_end
  - head_midpoint
  - chin #for rearing animals to properly track head movement
  - throat #same as above
  - tl_corner #top left corner to draw cage layout in R 
  - tr_corner #top right 
  - bl_corner #bottom left
  - br_corner #bottom right 
  - petri_12 #12 o'clock position of the petri dish. Indicated with 3 dots on the rim 
  - petri_03 # 3 o'clock position of the petri dish. Relative to the previous 12 o'clock position 
  - petri_06
  - petri_09
  - petri_12_3 # Halfway between the 12 o'clock and 3 o'clock position
  - petri_3_6
  - petri_6_9
  - petri_9_12

    # Fraction of video to start/stop when extracting frames for labeling/refinement

    # Fraction of video to start/stop when extracting frames for labeling/refinement

    # Fraction of video to start/stop when extracting frames for labeling/refinement

    # Fraction of video to start/stop when extracting frames for labeling/refinement


# Fraction of video to start/stop when extracting frames for labeling/refinement


# Fraction of video to start/stop when extracting frames for labeling/refinement


# Fraction of video to start/stop when extracting frames for labeling/refinement
  unique_bodyparts: []
  individuals:
  - animal
  with_identity:
method: bu
model:
  backbone:
    type: ResNet
    model_name: resnet50_gn
    output_stride: 16
    freeze_bn_stats: false
    freeze_bn_weights: false
  backbone_output_channels: 2048
  heads:
    bodypart:
      type: HeatmapHead
      weight_init: normal
      predictor:
        type: HeatmapPredictor
        apply_sigmoid: false
        clip_scores: true
        location_refinement: true
        locref_std: 7.2801
      target_generator:
        type: HeatmapGaussianGenerator
        num_heatmaps: 41
        pos_dist_thresh: 17
        heatmap_mode: KEYPOINT
        gradient_masking: false
        generate_locref: true
        locref_std: 7.2801
      criterion:
        heatmap:
          type: WeightedMSECriterion
          weight: 1.0
        locref:
          type: WeightedHuberCriterion
          weight: 0.05
      heatmap_config:
        channels:
        - 2048
        - 41
        kernel_size:
        - 3
        strides:
        - 2
      locref_config:
        channels:
        - 2048
        - 82
        kernel_size:
        - 3
        strides:
        - 2
net_type: resnet_50
runner:
  type: PoseTrainingRunner
  gpus:
  key_metric: test.mAP
  key_metric_asc: true
  eval_interval: 10
  optimizer:
    type: AdamW
    params:
      lr: 0.001
  scheduler:
    type: LRListScheduler
    params:
      lr_list:
      - - 0.0001
      - - 1e-05
      milestones:
      - 90
      - 120
  snapshots:
    max_snapshots: 5
    save_epochs: 25
    save_optimizer_state: false
train_settings:
  batch_size: 8
  dataloader_workers: 0
  dataloader_pin_memory: false
  display_iters: 500
  epochs: 200
  seed: 42

```

And here the code I used to create the trainin set and train the model:
```python
deeplabcut.create_training_dataset(
    config_file_path,
    net_type='resnet_50',
    augmenter_type='albumentations'
    )

deeplabcut.train_network(
    config_file_path,
    shuffle=1,
    max_snapshots_to_keep=20,
    displayiters=1000,
    #saveiters=5000
    epochs=500,
    save_epochs=10,
    
    )
```

**Comment by n-poulsen** [comment]
Thanks @tsievert - could I get a little more information from your project?

1. How big are your images?
2. Do all images have the same size?
3. Are there bodyparts that often aren't labeled?

As a first debug measure, could you try changing the data augmentation in your `pytorch_config.yaml` file to be:

```yaml
data:
  colormode: RGB
  inference:
    normalize_images: true
  train:
    affine:
      p: 0.5
      rotation: 30
      scaling:
      - 0.5
      - 1.25
      translation: 0
    gaussian_noise: 12.75
    normalize_images: true
```

And let me know if the model starts learning? Only training for ~20 epochs or so should be enough to see if the loss is going down.

**Comment by tsievert** [comment]
Hey @n-poulsen,
1. All picture are 1440 x 1080 pixels, and ~950 - 1000 KB
2. All frames share the same size
3. I have a set of bodyparts that occurs only in about 50% of frames, and two bodyparts that occur pretty rarely only. The rest appears often.

Output for a few epochs with the adjusted `pytorch_config.yaml`:
```bash
Using 2372 images and 125 for testing

Starting pose model training...
--------------------------------------------------
Epoch 1/20 (lr=0.001), train loss 0.01463
Epoch 2/20 (lr=0.001), train loss 0.01462
Epoch 3/20 (lr=0.001), train loss 0.01462
Epoch 4/20 (lr=0.001), train loss 0.01462
Epoch 5/20 (lr=0.001), train loss 0.01462
Epoch 6/20 (lr=0.001), train loss 0.01462
Epoch 7/20 (lr=0.001), train loss 0.01462
Epoch 8/20 (lr=0.001), train loss 0.01462
Epoch 9/20 (lr=0.001), train loss 0.01462
Training for epoch 10 done, starting evaluation
Epoch 10/20 (lr=0.001), train loss 0.01462, valid loss 0.01462
Model performance:
  metrics/test.rmse:         930.17
  metrics/test.rmse_pcutoff:    nan
  metrics/test.mAP:            0.00
  metrics/test.mAR:            0.00
```

And here the evaluation of the TF model trained with the same data (after 1.3 million iterations): 
```bash
With pcutoff of 0.8  train error: 4.55 pixels. Test error: 6.69 pixels
```

**Comment by n-poulsen** [comment]
Thanks for the info @tsievert! The loss really isn't moving at all, which is very weird - I haven't seen that issue exactly, and for the loss not to move at all it feels like the weights aren't being updated. Could you set `displayiters=1` during training to see if the loss changes between iterations (running for 20 iterations is enough, I just want to see if the loss is `0.01462` for each batch)

**Comment by tsievert** [comment]
Hey @n-poulsen,
I misread 20 iterations as 20 epochs, so here is the output in the middle of epoch 4. 

[R_7291685_DLC_train_model.txt](https://github.com/user-attachments/files/19319140/R_7291685_DLC_train_model.txt)

**Comment by n-poulsen** [comment]
Thanks @tsievert - it seems that the loss is at least changing a bit between batches. Could you try lowering the learning rate and seeing if that impacts learning? You can do that by setting:

```yaml
runner:
  ...
  optimizer:
    type: AdamW
    params:
      lr: 1e-04
```

And would you also be able to check whether the data you're loading contains valid entries? You can just report what's printed if you run the code:

```python
import deeplabcut.pose_estimation_pytorch as dlc_torch

loader = dlc_torch.DLCLoader("/path/to/project/config.yaml", shuffle=1)
print(f"Training data: {len(loader.df_train)}")
print(loader.df_train)

print(f"Test data: {len(loader.df_test)}")
print(loader.df_test)
```

**Comment by tsievert** [comment]
Output from reduced learning rate:
```pycon
Starting pose model training...
--------------------------------------------------
Epoch 1/500 (lr=0.0001), train loss 0.01239
Epoch 2/500 (lr=0.0001), train loss 0.00901
Epoch 3/500 (lr=0.0001), train loss 0.00792
Epoch 4/500 (lr=0.0001), train loss 0.00722
Epoch 5/500 (lr=0.0001), train loss 0.00686
Epoch 6/500 (lr=0.0001), train loss 0.00652
Epoch 7/500 (lr=0.0001), train loss 0.00624
Epoch 8/500 (lr=0.0001), train loss 0.00602
Epoch 9/500 (lr=0.0001), train loss 0.00582
Training for epoch 10 done, starting evaluation
Epoch 10/500 (lr=0.0001), train loss 0.00569, valid loss 0.00556
Model performance:
  metrics/test.rmse:          77.36
  metrics/test.rmse_pcutoff:  11.24
  metrics/test.mAP:           74.18
  metrics/test.mAR:           78.96
```
That looks more normal.


Output from other debug code:
```pycon
Loading DLC 3.0.0rc7...
DLC loaded in light mode; you cannot use any GUI (labeling, relabeling and standalone GUI)
Training data: 2372
scorer                                                                             Sievert  ...            
bodyparts                                                                             nose  ...  petri_9_12
coords                                                                                   x  ...           y
labeled-data Basler_acA1440-220um__40264738__20230912_112719915 img042993.png   821.360569  ...         NaN
             2023-11-03_ID075_BWm                               img208960.png          NaN  ...  765.394595
             2023-10-25_ID049_CTRL_preexposure                  img151820.png   449.380653  ...         NaN
             2024-07-22_ID236_DipropGlucol                      img059160.png   939.392437  ...  648.215237
             2023-09-26_ID066_BWm_preexposure                   img104350.png  1115.674122  ...         NaN
...                                                                                    ...  ...         ...
             2023-11-03_ID099_BWm_preexposure                   img011690.png          NaN  ...         NaN
             2023-10-24_ID095_POf                               img009940.png          NaN  ...         NaN
             2024-07-23_ID229_Eug                               img057170.png   194.737441  ...  581.562054
             2023-10-03_ID077_CTRL_preexposure                  img223430.png          NaN  ...         NaN
             2024-08-01_ID179_POf_preexposure                   img050560.png   407.819526  ...         NaN

[2372 rows x 82 columns]
Test data: 125
scorer                                                           Sievert  ...            
bodyparts                                                           nose  ...  petri_9_12
coords                                                                 x  ...           y
labeled-data 2023-09-20_ID082_POm             img013490.png   306.929787  ...  537.565221
             2023-10-30_ID068_BWf             img070530.png   805.173204  ...         NaN
             2023-11-01_ID068_POf             img217370.png  1127.638117  ...  536.643309
             2023-09-28_ID070_POm_preexposure img016277.png   344.095866  ...         NaN
             2023-10-31_ID068_POm             img089450.png   663.198285  ...  668.140756
...                                                                  ...  ...         ...
             2023-10-25_ID046_CTRL            img202210.png   828.675250  ...  353.655282
             2023-11-06_ID094_POm_preexposure img045450.png   732.007719  ...         NaN
             2023-10-04_ID039_BWm             img150230.png   266.263842  ...  642.157963
             2023-10-06_ID037_POm_preexposure img040740.png  1180.904125  ...         NaN
             2023-11-03_ID100_BWm             img122170.png          NaN  ...  428.138042

[125 rows x 82 columns]
```

**Comment by n-poulsen** [comment]
So the default learning rate was too aggressive for your dataset - I hadn't seen that yet but it may motivate me to modify the default (models should really converge on any well-labeled projects with default augmentation/learning parameters).

Thanks for reporting this @tsievert - it really helps to get bug reports when things aren't working on other datasets. I'll run some more tests on my end and see if the learning rate should be reduced. Let me know if you have other issues with convergence!

**Comment by tsievert** [comment]
Happy to help!
I'm quite impressed by the pytorch performance. By epoch 40 it's outperforming the TF model I had trained with the same data and 1.3 million iterations.

**Comment by MMathisLab** [comment]
thanks @tsievert for reporting, and glad it's working now! 💪
